python/ray/includes/object_ref.pxi

```diff
@@ -71,30 +92,27 @@ cdef class ObjectRef(BaseID):
         return self.as_future().__await__()
 
     def as_future(self):
+        """Wrap ObjectRef into an asyncio.Future object."""
         loop = asyncio.get_event_loop()
         py_future = loop.create_future()
 
         def callback(result):
             loop = py_future._loop
+            loop.call_soon_threadsafe(
+                functools.partial(_set_future_helper, py_future, result))
 
-            def set_future():
-                # Issue #11030, #8841
-                # If this future has result set already, we just need to
-                # skip the set result/exception procedure.
-                if py_future.done():
-                    return
+        self._on_completed(callback)
 
-                if isinstance(result, RayTaskError):
-                    py_future.set_exception(result.as_instanceof_cause())
-                elif isinstance(result, RayError):
-                    # Directly raise exception for RayActorError
-                    py_future.set_exception(result)
-                else:
-                    py_future.set_result(result)
+        # A hack to keep a reference to the object ref for ref counting.
+        py_future.object_ref = self
+        return py_future
 
-            loop.call_soon_threadsafe(set_future)
+    def as_concurrent_future(self) -> concurrent.futures.Future:
+        """Wrap ObjectRef into a concurrent.futures.Future"""
```

Comment:
Please add information to the docstring about the behavior and compatibility with concurrent futures (what methods we do and do not support). It would also be great if you could add the same to the asyncio version.